src/librustc_trans/trans/callee.rs

```diff
@@ -603,11 +603,27 @@ pub fn trans_method_call<'a, 'blk, 'tcx>(bcx: Block<'blk, 'tcx>,
     let _icx = push_ctxt("trans_method_call");
     debug!("trans_method_call(call_expr={})", call_expr.repr(bcx.tcx()));
     let method_call = MethodCall::expr(call_expr.id);
-    let method_ty = (*bcx.tcx().method_map.borrow())[method_call].ty;
+    let (origin, method_ty) =
+        bcx.tcx().method_map
+                 .borrow()
+                 .get(&method_call)
+                 .map(|method| (method.origin.clone(), method.ty))
+                 .unwrap();
+    let method_ty = match origin {
+        ty::MethodTraitObject(_) => match method_ty.sty {
+            ty::ty_bare_fn(_, ref fty) => {
+                let fty = meth::opaque_method_ty(bcx.tcx(), fty);
+                ty::mk_bare_fn(bcx.tcx(), None, bcx.tcx().mk_bare_fn(fty))
+            }
+            _ => method_ty
+        },
+        _ => method_ty
+    };
+    let method_ty = common::monomorphize_type(bcx, method_ty);
```

Comment:
Does `bcx` not have a method for this? `common` is rather jarring IMO (though I dislike the module more than the name).